crate_universe/extension.bzl

```diff
@@ -0,0 +1,224 @@
+
+    For an example json file, see examples/bzlmod/hello_world/annotations.json.
+    """
+    annotations = {}
+
+    for file in annotation_files:
+        content = json.decode(module_ctx.read(file))
+        for name, annotations_for_crate in content.items():
+            for annotation in annotations_for_crate:
+                annotation = _crate_universe_crate.annotation(**annotation)
+                annotations.setdefault(name, [])
+                annotations[name].append(annotation)
+
+    return annotations
+
+def _generate_hub_and_spokes(module_ctx, cargo_bazel, mod_path, cfg):
+    annotations = _generate_annotations(module_ctx, cfg.annotation_files)
+    tag_path = mod_path
+    repo_name = "crates"
+    if cfg.suffix:
+        tag_path = mod_path.get_child(cfg.suffix)
+        repo_name += "_" + cfg.suffix
+
+    cargo_lockfile = module_ctx.path(cfg.cargo_lockfile)
+
+    rendering_config = json.decode(render_config(
+        regen_command = "Run 'cargo update [--workspace]'",
+    ))
+    config_file = tag_path.get_child("config.json")
+    module_ctx.file(
+        config_file,
+        executable = False,
+        content = generate_config_file(
+            module_ctx,
+            mode = "remote",
+            annotations = annotations,
+            generate_build_scripts = cfg.generate_build_scripts,
+            supported_platform_triples = cfg.supported_platform_triples,
+            generate_target_compatible_with = True,
+            repository_name = repo_name,
+            output_pkg = repo_name,
+            workspace_name = repo_name,
+            generate_binaries = cfg.generate_binaries,
+            rendering_config = rendering_config,
+        ),
+    )
+
+    manifests = {module_ctx.path(m): m for m in cfg.manifests}
+    splicing_manifest = tag_path.get_child("splicing_manifest.json")
+    module_ctx.file(
+        splicing_manifest,
+        executable = True,
+        content = generate_splicing_manifest(
+            packages = {},
+            splicing_config = "",
+            cargo_config = cfg.cargo_config,
+            manifests = {str(k): str(v) for k, v in manifests.items()},
+            manifest_to_path = module_ctx.path,
+        ),
+    )
+
+    splicing_output_dir = tag_path.get_child("splicing-output")
+    cargo_bazel([
+        "splice",
+        "--output-dir",
+        splicing_output_dir,
+        "--config",
+        config_file,
+        "--splicing-manifest",
+        splicing_manifest,
+        "--cargo-lockfile",
+        cargo_lockfile,
+    ])
+
+    # Create a lockfile, since we need to parse it to generate spoke
+    # repos.
+    lockfile_path = tag_path.get_child("lockfile.json")
+    module_ctx.file(lockfile_path, "")
+
+    # cargo-bazel generate takes the lockfile as input, but also writes
+    # to the lockfile. This means that even though nothing changes, the
+    # modified timestamp of the file is updated. Since the lock file is
```

Comment:
https://github.com/bazelbuild/rules_rust/pull/2386 should hopefully allow removing this?